### PR TITLE
Add an InMemoryExporter<T>

### LINKED ISSUE #1334: In Memory Exporter
# Feature Request

This is required by the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-guidelines.md#requirements).
Currently we do have something similar in the test cases. We probably want to make an `InMemoryExporter<T>` and leverage it in the test cases (and folks who use OpenTelemetry can also leverage the InMemoryExporter to test their instrumentation / integration).

### PR DESCRIPTION
**Pull request by reyang**
Fixes #1334.

## Changes

* Added an in-memory exporter according to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-guidelines.md#requirements).

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    public class InMemoryExporter<T> : BaseExporter<T>
+        where T : class
+    {
+        private readonly IList<T> trait;
+
+        public InMemoryExporter(InMemoryExporterOptions options)
+        {
+            if (options.Trait == null)
+            {
+                throw new ArgumentNullException("nameof(options.Trait)");
```
Comment (eddynaka) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:32]:
```suggestion
                throw new ArgumentNullException(nameof(options.Trait));
```

File: src/OpenTelemetry.Exporter.InMemory/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+```shell
+dotnet add package OpenTelemetry.Exporter.InMemory
+```
+
+## Configuration
+
+```csharp
+var list = new List<Activity>();
+
+using Sdk.CreateTracerProviderBuilder()
```
Comment (eddynaka) [inline src/OpenTelemetry.Exporter.InMemory/README.md:22]: nit: add an example for the logger extension

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs (hunk 1 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:32]: Thanks for the catch!

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporterHelperExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            var options = new InMemoryExporterOptions();
+            configure?.Invoke(options);
+            return builder.AddProcessor(new SimpleExportProcessor<Activity>(new InMemoryExporter<Activity>(options)));
+        }
+
+#if NETSTANDARD2_0
```
Comment (eddynaka) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporterHelperExtensions.cs:48]: what versions of the framework are we going to support? if we have net48, this condition wouldn't work i think.

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Exporter
+{
+    public class InMemoryExporterOptions
+    {
+        public IList<object> Trait;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs:23]: Public field! The horror! Should have `{ get; set; }` 😄 Might even want:
```csharp
public IList<object> Trait { get; set; } = new List<object>();
```
Not sure if you wanted it null by default?

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            }
+
+            this.trait = (IList<T>)options.Trait;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:35]: nit: Doesn't look like the cast is needed.

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public override ExportResult Export(in Batch<T> batch)
+        {
+            foreach (var data in batch)
+            {
+                this.trait.Add(data);
+            }
```
Comment (ThomsonTan) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:43]: Do we need dump the in-memory data to console for test purpose?

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporterHelperExtensions.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporterHelperExtensions.cs:48]: This library only targets the following frameworks:
```xml
<TargetFrameworks>net452;net46;netstandard2.0</TargetFrameworks>
```

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs (hunk 4 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs]: Oops, you're right, fixed.

Yes I put null/no-op by default.

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs (hunk 6 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:38]: Nope.

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+using OpenTelemetry;
+
+namespace OpenTelemetry.Exporter
+{
+    public class InMemoryExporter<T> : BaseExporter<T>
+        where T : class
+    {
+        private readonly ICollection<object> trait;
+
+        public InMemoryExporter(InMemoryExporterOptions options = null)
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:28]: You could do InMemoryExpoterOptions<T> here so the list/collection could be strongly typed (ICollection<T> Collection { get; set; }).

Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:28]: I did this intentional - trying to get feedback: in this way we can have a single collection used by different types of exporters, I've put a demo in the README.md. Might be over-enginering?

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Exporter
+{
+    public class InMemoryExporterOptions
+    {
+        public ICollection<object> Collection { get; set; }
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs:23]: `Collection` is a little generic, maybe we should use something like `ExportedItems`?

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs (hunk 7 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:28]: FYI, this is what I was trying:


```csharp
var list = new List<object>();
var activityExporter = new InMemoryExporter<Activity>(new InMemoryExporterOptions { Collection = list });
var logExporter = new InMemoryExporter<LogRecord>(new InMemoryExporterOptions { Collection = list });
// do something about the synthesized data.
```

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs (hunk 8 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporterOptions.cs:23]: That's a much better name 👍

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs (hunk 7 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:28]: Ah, cool. I think a typed version would be more useful, more often, but given this is for unit testing primarily (?), a cast is probably fine.

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        private readonly ICollection<object> exportedItems;
+
+        public InMemoryExporter(InMemoryExporterOptions options = null)
+        {
+            this.exportedItems = options?.ExportedItems;
+        }
+
+        public override ExportResult Export(in Batch<T> batch)
+        {
+            foreach (var data in batch)
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:35]: Prob faster to do the null check outside of the loop:
```csharp
if (this.exportedItems != null)
{
   foreach (var data in batch)
   {
      this.exportedItems.Add(data);
   }
}
```

Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs:35]: Nice catch.

I've changed it to return `Export.Failure` when `exportedItems` is null.

File: src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs (hunk 7 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs]: I've added contravariance, see if you like it @CodeBlanch.

**Pull request merged**